### ISSUE TITLE
Because of a timeout, Catalyst uses the limit price rather than the fill price to measure performance.

### ISSUE THREAD

**Issue by sam31415** [enigmampc/catalyst#321]
Dear Catalyst Maintainers,

Before I tell you about my issue, let me describe my environment:

# Environment

* Operating System: Linux trading-bot 4.13.0-39-generic #44-Ubuntu SMP Thu Apr 5 14:25:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
* Python Version: Python 3.6.4 :: Anaconda, Inc.
* Python Bitness: 64
* How did you install Catalyst: conda
* Python packages: 
aiodns                    1.1.1                     <pip>                      
aiohttp                   3.1.3                     <pip>                      
alembic                   0.9.9                     <pip>                      
async-timeout             2.0.1                     <pip>                      
attrdict                  2.0.0                     <pip>                      
attrs                     17.4.0                    <pip>                      
bcolz                     0.12.1                    <pip>                      
bcolz                     1.1.2            py36h00f5784_0                      
bokeh                     0.12.14                  py36_0                      
boto3                     1.7.4                     <pip>                      
botocore                  1.10.4                    <pip>                      
Bottleneck                1.2.1                     <pip>                      
bzip2                     1.0.6                h9a117a8_4                      
ca-certificates           2017.08.26           h1d4fec5_0                      
cchardet                  2.1.1                     <pip>                      
ccxt                      1.13.1                    <pip> 
certifi                   2018.1.18                py36_0                                                                                              
certifi                   2018.4.16                 <pip>                      
chardet                   3.0.4                     <pip>                      
click                     6.7              py36h5253387_0                      
cloudpickle               0.5.2                    py36_1                      
configparser              3.5.0                     <pip>                      
contextlib2               0.5.5                     <pip>                      
cycler                    0.10.0                    <pip>                      
cyordereddict             1.0.0                     <pip>                      
Cython                    0.28.2                    <pip>                      
cython                    0.27.3           py36h1860423_0                      
cytoolz                   0.9.0.1          py36h14c3975_0                      
dask                      0.17.2                   py36_0                      
dask-core                 0.17.2                   py36_0                      
decorator                 4.3.0                     <pip>                      
distributed               1.21.4                   py36_0                      
docutils                  0.14                      <pip>                      
empyrical                 0.2.1                     <pip>                      
enigma-catalyst           0.5.9                     <pip>                      
eth-abi                   1.0.0                     <pip>                      
eth-account               0.1.0a2                   <pip>                      
eth-hash                  0.1.2                     <pip>                      
eth-keyfile               0.5.1                     <pip>                      
eth-keys                  0.2.0b3                   <pip>                      
eth-rlp                   0.1.0                     <pip>                      
eth-utils                 1.0.2                     <pip>                      
hdf5                      1.10.1               h9caa474_1                      
heapdict                  1.0.0                    py36_2                      
hexbytes                  0.1.0                     <pip>                      
idna                      2.6                       <pip>                      
idna-ssl                  1.0.1                     <pip>                      
intel-openmp              2018.0.0                      8                      
intervaltree              2.1.0                     <pip>                      
jinja2                    2.10             py36ha16c418_0                      
jmespath                  0.9.3                     <pip>                      
kiwisolver                1.0.1                     <pip>                      
libedit                   3.1                  heed3624_0                      
libffi                    3.2.1                hd88cf55_4
libgcc-ng                 7.2.0                hdf63c60_3                                                                                              [38/577]
libgfortran-ng            7.2.0                hdf63c60_3                      
libstdcxx-ng              7.2.0                hdf63c60_3                      
locket                    0.2.0            py36h787c0ad_1                      
Logbook                   1.3.3                     <pip>                      
lru-dict                  1.1.6                     <pip>                      
lxml                      4.2.1                     <pip>                      
lzo                       2.10                 h49e0be7_2                      
Mako                      1.0.7                     <pip>                      
markupsafe                1.0              py36hd9260cd_1                      
matplotlib                2.2.2                     <pip>                      
mkl                       2018.0.2                      1                      
mpld3                     0.3                       <pip>                      
msgpack-python            0.5.5            py36h6bb024c_0                      
multidict                 4.2.0                     <pip>                      
multipledispatch          0.5.0                     <pip>                      
ncurses                   6.0                  h9df7e31_2                      
networkx                  2.1                       <pip>                      
numexpr                   2.6.4            py36hc4a3f9a_0                      
numpy                     1.14.2                    <pip>                      
numpy                     1.14.2           py36hdbf6ddf_0                      
openssl                   1.0.2n               hb7f436b_0                      
packaging                 17.1                     py36_0                      
pandas                    0.22.0           py36hf484d3e_0                      
pandas                    0.19.2                    <pip>                      
pandas-datareader         0.6.0                     <pip>                      
partd                     0.3.8            py36h36fd896_0                      
patsy                     0.5.0                     <pip>                      
patsy                     0.5.0                    py36_0                      
pip                       10.0.0                    <pip>                      
pip                       9.0.1                    py36_5                      
psutil                    5.4.3            py36h14c3975_0                      
pycares                   2.3.0                     <pip>                      
pycryptodome              3.6.1                     <pip>                      
pyparsing                 2.2.0            py36hee85983_1                      
pytables                  3.4.2            py36h3b5282a_2                      
python                    3.6.4                hc3d631a_3                      
python-dateutil           2.7.2                     <pip>
python-dateutil           2.7.0                    py36_0  
python-editor             1.0.3                     <pip>
pytz                      2018.3                   py36_0  
pytz                      2018.4                    <pip>
pyyaml                    3.12             py36hafb9ca4_1  
readline                  7.0                  ha6073c6_4  
redo                      1.6                       <pip>
requests                  2.18.4                    <pip>
requests-file             1.4.3                     <pip>
requests-ftp              0.3.1                     <pip>
requests-toolbelt         0.8.0                     <pip>
rlp                       0.6.0                     <pip>
s3transfer                0.1.13                    <pip>
scipy                     1.0.0            py36hbf646e7_0  
scipy                     1.0.1                     <pip>
setuptools                38.5.1                   py36_0  
setuptools                39.0.1                    <pip>
six                       1.10.0                    <pip>
six                       1.11.0           py36h372c433_1  
sortedcontainers          1.5.9                    py36_0  
SQLAlchemy                1.2.6                     <pip>
sqlite                    3.22.0               h1bed415_0  
statsmodels               0.8.0                     <pip>
statsmodels               0.8.0            py36h8533d0b_0  
tabulate                  0.8.2                     <pip>
tblib                     1.3.2            py36h34cf8b6_0  
tk                        8.6.7                hc745277_3  
toolz                     0.9.0                    py36_0  
tornado                   5.0                      py36_0  
urllib3                   1.22                      <pip>
web3                      4.1.0                     <pip>
websockets                4.0.1                     <pip>
wheel                     0.30.0           py36hfd4bba0_1  
wrapt                     1.10.11                   <pip>
xz                        5.2.3                h55aa19d_2  
yaml                      0.1.7                had09818_2  
yarl                      1.1.0                     <pip>
zict                      0.1.3            py36h3a3bf81_0  
zlib                      1.2.11               ha838bed_2

# Description of Issue

This issue may be related to #315. The algorithm sent an order to Binance to buy ven_btc with a limit at 120% the current price (to make it effectively a market order), using the `order`. Here is the log:
```
[2018-05-03 08:10:21.484622] INFO: CCXT: order amount adjusted by binance from 3.39658668 to 3.0                                                               
[2018-05-03 08:10:21.485248] NOTICE: CryptoTrendIII-Binance: Ordering 0.0593 of ven.                                                                                                                                                            
[2018-05-03 08:11:27.866741] WARNING: CCXT: unable to fetch trades binance / ven_btc: binance https://api.binance.com/api/v3/myTrades?timestamp=1525335087783&r
ecvWindow=5000&symbol=VENBTC&limit=100&signature=27b1004b00a91a6aee797a38c1d73d2af82133d703cdf7b90d866bf335998c50 GET ('Connection aborted.', ConnectionResetEr
ror(104, 'Connection reset by peer')) {"symbol":"BTGBTC","priceChange":"0.00008500","priceChangePercent":"1.066","weightedAvgPrice":"0.00818026","prevClosePric
e":"0.00797400","lastPrice":"0.00805500","lastQty":"6.25000000","bidPrice":"0.00804100","bidQty":"10.84000000","askPrice":"0.00805500","askQty":"1.52000000","o
penPrice":"0.00797000","highPrice":"0.00840000","lowPrice":"0.00792000","volume":"110402.19000000","quoteVolume":"903.11877244","openTime":1525248656136,"close
Time":1525335056136,"firstId":2752248,"lastId":2785236,"count":32989}
[2018-05-03 08:11:27.867408] WARNING: CCXT: unable to fetch account trades, trying an alternate method to find executed order 26747514 / ven_btc: Request faile
d: binance https://api.binance.com/api/v3/myTrades?timestamp=1525335087783&recvWindow=5000&symbol=VENBTC&limit=100&signature=27b1004b00a91a6aee797a38c1d73d2af8
2133d703cdf7b90d866bf335998c50 GET ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer')) {"symbol":"BTGBTC","priceChange":"0.00008500"
,"priceChangePercent":"1.066","weightedAvgPrice":"0.00818026","prevClosePrice":"0.00797400","lastPrice":"0.00805500","lastQty":"6.25000000","bidPrice":"0.00804
100","bidQty":"10.84000000","askPrice":"0.00805500","askQty":"1.52000000","openPrice":"0.00797000","highPrice":"0.00840000","lowPrice":"0.00792000","volume":"1
10402.19000000","quoteVolume":"903.11877244","openTime":1525248656136,"closeTime":1525335056136,"firstId":2752248,"lastId":2785236,"count":32989}
[2018-05-03 08:11:29.198010] INFO: CCXT: order 26747514 executed in full: 3.0 ven_btc
[2018-05-03 08:11:29.199143] INFO: exchange_blotter: filled order 26747514 / ven_btc: 3.0, avg price: 0.00067446
```
On the exchange I observe this:
`2018-05-03 10:10:21 | VEN/BTC | Limit | Buy | 0.00056250 | 0.00067446 | 3 | 3 | 0.00168750 BTC | -- | Filled`
The filling price is 0.00056250 and the limit price is 0.00067446.

Now remark that Catalyst recognized the order as filled, but at the limit price (0.00067446), rather than at the fill price (0.00056250). This results in an apparent loss.

Remarks:
- This is due to the timeout, because numerous similar other orders were filled without problem. Hopefully the fix of #315 will also fix this.
- I am using limit order with a high limit price rather than a market order because I am also testing this algorithm on Poloniex and Poloniex does not accept market orders.
- The exact same problem also occurred a while ago on Poloniex, but I don't have the logs available.
- It would be great if someone could answer #312. Because when a problem like this arises, the options are either to erase the algorithm history and start again with a clean sheet or to continue trading with flawed performance reports including fictive losses. In the present case, I would need to modify the "cost_basis" attribute of the ven_btc position object, to replace the limit price by the fill price.

Sincerely,
Samuel

**Comment by sam31415** [comment]
Here is the same problem when selling ltc_btc on Poloniex (again using the `order` function with a limit markedly lower than the current price to simulate a market order):
```
[2018-05-03 10:01:10.809659] WARNING: CCXT: unable to fetch trades poloniex / ltc_btc: poloniex POST https://poloniex.com/tradingApi HTTPSConnectionPool(host='poloniex.com', port=443): Read timed out. (read timeout=10)
[2018-05-03 10:01:10.810143] WARNING: CCXT: unable to fetch account trades, trying an alternate method to find executed order 171103384391 / ltc_btc: Request failed: poloniex POST https://poloniex.com/tradingApi HTTPSConnectionPool(host='poloniex.com', port=443): Read timed out. (read timeout=10)
[2018-05-03 10:01:11.040339] INFO: CCXT: order 171103384391 executed in full: 0.16344431 ltc_btc
[2018-05-03 10:01:11.041108] INFO: exchange_blotter: filled order 171103384391 / ltc_btc: -0.16344431, avg price: 0.014771745
```
If I look at the exchange, here is what I see:
`
LTC/BTC    Sell    Exchange    0.01641305    0.16344431    0.00000671 BTC (0.25%)    0.00267590 BTC    2018-05-03 10:00:11
`

The sell ltc order was filled at 0.01641305, but Catalyst recorded a sale at the limit price 0.014771745.

**Comment by sam31415** [comment]
And again:
```
[2018-05-03 20:01:18.271678] WARNING: CCXT: unable to fetch trades poloniex / etc_btc: poloniex POST https://poloniex.com/tradingApi HTTPSConnectionPool(host='poloniex.com', port=443): Read timed out. (read timeout=10)          
[2018-05-03 20:01:18.272127] WARNING: CCXT: unable to fetch account trades, trying an alternate method to find executed order 207668955339 / etc_btc: Request failed: poloniex POST https://poloniex.com/tradingApi HTTPSConnectionPool(host='poloniex.com', port=443): Read timed out. (read timeout=10)                                             
[2018-05-03 20:01:18.501654] INFO: CCXT: order 207668955339 executed in full: 1.05553142 etc_btc                  
[2018-05-03 20:01:18.502480] INFO: exchange_blotter: filled order 207668955339 / etc_btc: -1.05553142, avg price: 0.002118159
```
But on Poloniex the orders were filled as follows:
```
  | ETC/BTC | Sell | Exchange | 0.00235253 | 0.26465471 | 0.00000156 BTC (0.25%) | 0.00062104 BTC | 2018-05-03 20:00:46
  | ETC/BTC | Sell | Exchange | 0.00235401 | 0.79087671 | 0.00000465 BTC (0.25%) | 0.00185708 BTC | 2018-05-03 20:00:46
```
Catalyst used the limit price (0.002118159) rather than the fill prices (0.00235401, 0.00235253).

**Comment by AvishaiW** [comment]
Hi, Thanks for noting this out! The problem was that once fetching the trades was rejected, the code moved on to get the info through the order data, which doesn't have any info regarding the actual price. Therefore, I raised an error instead, leading the code to retry the trade fetching several times. Hope this will work out, still needed to be tested.

**Comment by lenak25** [comment]
Fixed at release 0.5.10.
